### PR TITLE
Add seedjob agent

### PR DESCRIPTION
**Pull request by jakalkhalili**

### CODE REVIEW

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go [hunk 1]
```diff
@@ -233,6 +261,123 @@ func (s *SeedJobs) isRecreatePodNeeded(jenkins v1alpha2.Jenkins) bool {
 	return false
 }
 
+// CreateAgent deploys Jenkins agent to Kubernetes cluster
+func CreateAgent(jenkinsClient jenkinsclient.Jenkins, k8sClient client.Client, jenkinsManifest *v1alpha2.Jenkins, namespace string, agentName string) error {
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: make it private `CreateAgent` -> `createAgent` and make part of SeedJobs struct

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go [hunk 2]
```diff
@@ -233,6 +261,123 @@ func (s *SeedJobs) isRecreatePodNeeded(jenkins v1alpha2.Jenkins) bool {
...
+
+	deployments := &apps.DeploymentList{}
+	exists = false
+	secret, err := jenkinsClient.GetNodeSecret(agentName)
+	if err != nil {
+		return err
+	}
+
+	deployment := agentDeployment(jenkinsManifest, namespace, agentName, secret)
+	err = k8sClient.List(context.TODO(), &client.ListOptions{}, deployments)
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: don't list deployments

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go [hunk 3]
```diff
@@ -233,6 +261,123 @@ func (s *SeedJobs) isRecreatePodNeeded(jenkins v1alpha2.Jenkins) bool {
...
+		for _, deployment := range deployments.Items {
+			if deployment.ObjectMeta.Name == fmt.Sprintf("%s-deployment", agentName) {
+				exists = true
+			}
+		}
+	}
+
+	// Create deployment if not exists
+	if !exists {
+		err = k8sClient.Create(context.TODO(), deployment)
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: always create if err is already exists update it

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go [hunk 4]
```diff
@@ -42,6 +46,12 @@ const (
 	// JenkinsCredentialTypeLabelName is label for kubernetes-credentials-provider-plugin which determine Jenkins
 	// credential type
 	JenkinsCredentialTypeLabelName = "jenkins.io/credentials-type"
+
+	// AgentName is the name of seed job agent
+	AgentName = "seed-job-agent"
+
+	// AgentNamespace is the namespace of seed job agent
+	AgentNamespace = "default"
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: remove it

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go [hunk 5]
```diff
@@ -88,6 +98,24 @@ func (s *SeedJobs) EnsureSeedJobs(jenkins *v1alpha2.Jenkins) (done bool, err err
...
 
+	if len(seedJobIDs) > 0 {
+		err := CreateAgent(s.jenkinsClient, s.k8sClient, jenkins, jenkins.Namespace, AgentName)
+		if err != nil {
+			panic(err)
+		}
+	} else if len(seedJobIDs) == 0 {
+		err := s.k8sClient.Delete(context.TODO(), &appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: AgentNamespace,
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: unhardcode it `Namespace: jenkins.Namespace`

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go [hunk 6]
```diff
@@ -233,6 +261,123 @@ func (s *SeedJobs) isRecreatePodNeeded(jenkins v1alpha2.Jenkins) bool {
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-deployment", agentName),
+			Namespace: namespace,
+		},
+		Spec: apps.DeploymentSpec{
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  fmt.Sprintf("%s-container", agentName),
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: Name: "jnlp"

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go [hunk 7]
```diff
@@ -233,6 +261,123 @@ func (s *SeedJobs) isRecreatePodNeeded(jenkins v1alpha2.Jenkins) bool {
...
+					},
+				},
+			},
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app": fmt.Sprintf("%s-selector", agentName),
+				},
+			},
+		},
+		Status: apps.DeploymentStatus{},
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: remove it, not needed

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go [hunk 8]
```diff
@@ -233,6 +261,123 @@ func (s *SeedJobs) isRecreatePodNeeded(jenkins v1alpha2.Jenkins) bool {
...
+	}
+}
+
 // seedJobConfigXML this is the XML representation of seed job
 var seedJobConfigXML = `
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: rewrite it to groovy script

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go [hunk 9]
```diff
@@ -67,40 +63,71 @@ func (s *SeedJobs) EnsureSeedJobs(jenkins *v1alpha2.Jenkins) (done bool, err err
...
+		if err != nil {
+			return false, err
+		}
+	} else if len(jenkins.Spec.SeedJobs) == 0 {
+		err := s.k8sClient.Delete(context.TODO(), &appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: jenkins.Namespace,
+				Name:      fmt.Sprintf("%s-deployment", AgentName),
+			},
+		})
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: apierrors.IsNotFound

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go [hunk 10]
```diff
@@ -67,40 +63,71 @@ func (s *SeedJobs) EnsureSeedJobs(jenkins *v1alpha2.Jenkins) (done bool, err err
...
 
 // createJob is responsible for creating jenkins job which configures jenkins seed jobs and deploy keys
-func (s *SeedJobs) createJob() error {
-	_, created, err := s.jenkinsClient.CreateOrUpdateJob(seedJobConfigXML, ConfigureSeedJobsName)
-	if err != nil {
-		return err
-	}
-	if created {
-		s.logger.Info(fmt.Sprintf("'%s' job has been created", ConfigureSeedJobsName))
+func (s *SeedJobs) createJobs(jenkins *v1alpha2.Jenkins) (bool, error) {
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: `(bool, error)` -> `(requeue bool, err error)`

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go [hunk 11]
```diff
@@ -233,56 +221,126 @@ func (s *SeedJobs) isRecreatePodNeeded(jenkins v1alpha2.Jenkins) bool {
...
-          <description></description>
-          <defaultValue>cicd/jobs/*.jenkins</defaultValue>
-          <trim>false</trim>
-        </hudson.model.StringParameterDefinition>
-      </parameterDefinitions>
-    </hudson.model.ParametersDefinitionProperty>
-  </properties>
-  <definition class="org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition" plugin="workflow-cps@2.61">
-    <script>
+// CreateAgent deploys Jenkins agent to Kubernetes cluster
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: `// CreateAgent deploys Jenkins agent to Kubernetes cluster` -> `// createAgent deploys Jenkins agent to Kubernetes cluster`

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go [hunk 12]
```diff
@@ -233,6 +261,123 @@ func (s *SeedJobs) isRecreatePodNeeded(jenkins v1alpha2.Jenkins) bool {
 	return false
 }
 
+// CreateAgent deploys Jenkins agent to Kubernetes cluster
+func CreateAgent(jenkinsClient jenkinsclient.Jenkins, k8sClient client.Client, jenkinsManifest *v1alpha2.Jenkins, namespace string, agentName string) error {
+	var exists bool
+
+	nodes, err := jenkinsClient.GetAllNodes()
```
[the remark this answers is not in the document]
Comment (tomaszsek) [reply pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: ???

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go [hunk 13]
```diff
@@ -233,56 +221,126 @@ func (s *SeedJobs) isRecreatePodNeeded(jenkins v1alpha2.Jenkins) bool {
...
+
+	secret, err := jenkinsClient.GetNodeSecret(agentName)
+	if err != nil {
+		return err
+	}
+
+	deployment := agentDeployment(jenkinsManifest, namespace, agentName, secret)
+
+	err = k8sClient.Create(context.TODO(), deployment)
+	if err != nil {
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: `if err != nil {` -> `if err != nil && apierrors.IsAlreadyExists(err) {`

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go [hunk 14]
```diff
@@ -233,56 +221,126 @@ func (s *SeedJobs) isRecreatePodNeeded(jenkins v1alpha2.Jenkins) bool {
...
+		err := k8sClient.Update(context.TODO(), deployment)
+		if err != nil {
+			return err
+		}
+	}
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: `}` -> 
```
} else if err != nil {
    return err
}
```

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go (hunk 6 above)
Comment (tomaszsek) [reply pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: ???

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs_test.go [hunk 15]
```diff
@@ -219,3 +165,41 @@ func TestSeedJobs_isRecreatePodNeeded(t *testing.T) {
...
+			},
+		})
+
+		assert.NoError(t, err)
+
+		// then
+		err = seedJobsClient.createAgent(jenkinsClient, fakeClient, jenkinsCustomResource(), namespace, agentName)
+		assert.NoError(t, err)
+	})
+}
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/user/seedjobs/seedjobs_test.go]: add test which will delete agent deployment when seed job list is empty

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go (hunk 9 above)
Comment (jakalkhalili) [reply pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: done??

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go (hunk 10 above)
Comment (jakalkhalili) [reply pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: done

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go (hunk 11 above)
Comment (jakalkhalili) [reply pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: done

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (jakalkhalili) [reply pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: done

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go (hunk 6 above)
Comment (jakalkhalili) [reply pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: done?

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go [hunk 16]
```diff
@@ -233,56 +222,115 @@ func (s *SeedJobs) isRecreatePodNeeded(jenkins v1alpha2.Jenkins) bool {
...
-    </hudson.model.ParametersDefinitionProperty>
-  </properties>
-  <definition class="org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition" plugin="workflow-cps@2.61">
-    <script>
+// createAgent deploys Jenkins agent to Kubernetes cluster
+func (s SeedJobs) createAgent(jenkinsClient jenkinsclient.Jenkins, k8sClient client.Client, jenkinsManifest *v1alpha2.Jenkins, namespace string, agentName string) error {
+	_, err := jenkinsClient.GetNode(agentName)
+
+	// Create node if not exists
+	if err != nil {
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: create agent if err tells agent not exists, else return err

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go (hunk 13 above)
Comment (tomaszsek) [reply pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: wrong line

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go (hunk 6 above)
Comment (tomaszsek) [reply pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: ???

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs_test.go (hunk 15 above)
Comment (tomaszsek) [reply pkg/controller/jenkins/configuration/user/seedjobs/seedjobs_test.go]: ???

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go (hunk 13 above)
Comment (jakalkhalili) [reply pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: done

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go (hunk 16 above)
Comment (jakalkhalili) [reply pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: done

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go (hunk 6 above)
Comment (jakalkhalili) [reply pkg/controller/jenkins/configuration/user/seedjobs/seedjobs.go]: done!

File: pkg/controller/jenkins/configuration/user/seedjobs/seedjobs_test.go (hunk 15 above)
Comment (jakalkhalili) [reply pkg/controller/jenkins/configuration/user/seedjobs/seedjobs_test.go]: done
